# 01. 프로젝트 개요 — Message Broker Comparison Lab

`난이도: 입문` · `소요: 20분` · `사전 지식: Python 기초, FastAPI 개념`

---

## 목차
1. [프로젝트 디렉토리 구조](#1.-프로젝트-디렉토리-구조)
2. [환경 설정 (sys.path)](#2.-환경-설정-(sys.path))
3. [AbstractBroker — 공통 인터페이스 분석](#3.-AbstractBroker---공통-인터페이스-분석)
4. [pydantic-settings 설정 분석](#4.-pydantic-settings-설정-분석)
5. [Pydantic 스키마 분석](#5.-Pydantic-스키마-분석)
6. [싱글턴 패턴과 Lifespan 관리](#6.-싱글턴-패턴과-Lifespan-관리)
7. [Docker 인프라 확인](#7.-Docker-인프라-확인)
8. [첫 메시지 발행 — 각 브로커 테스트](#8.-첫-메시지-발행---각-브로커-테스트)
9. [Swagger UI 활용법](#9.-Swagger-UI-활용법)
10. [정리](#10.-정리---핵심-포인트)

---

### 학습 목표
이 노트북을 완료하면 다음을 이해할 수 있습니다:
- 프로젝트의 전체 구조와 각 모듈의 역할
- ABC 패턴으로 브로커 인터페이스를 통일하는 방법
- Docker 인프라 구성과 헬스체크
- 세 브로커(Redis, RabbitMQ, Kafka)에 메시지를 발행하는 기본 흐름

### 사전 준비
```bash
docker compose up -d   # 인프라 실행
uv run python main.py  # FastAPI 서버 실행 (별도 터미널)
```

> **Tip** — Jupyter에서 `await` 구문이 바로 동작합니다. `asyncio.run()`으로 감쌀 필요가 없습니다.

---

#### 노트북 네비게이션
| 이전 | 현재 | 다음 |
|------|------|------|
| — | **01. 프로젝트 개요** | [02. Redis Deep Dive](./02_redis_deep_dive.ipynb) |

> **학습 경로**: `01 개요` > `02 Redis` · `03 RabbitMQ` · `04 Kafka` > `05 벤치마크` > `06 모니터링` > `07 고급패턴` > `08 안정성` > `09 동시성` > `10 Saga`

---
## 1. 프로젝트 디렉토리 구조

```sql
kafkaRabbitTest/
├── main.py                    # FastAPI 진입점 (uvicorn 실행)
├── docker-compose.yml         # Redis + RabbitMQ + Kafka + 모니터링
├── app/
│   ├── config.py              # pydantic-settings 설정 (환경변수 → 타입 안전 객체)
│   ├── schemas.py             # Pydantic 요청/응답 모델 (SRP)
│   ├── lifespan.py            # FastAPI 라이프사이클 (브로커 connect/disconnect)
│   ├── brokers/
│   │   ├── base.py            # AbstractBroker - 공통 계약 (Strategy Pattern)
│   │   ├── redis_broker.py    # Redis 구현 (Pub/Sub, Stream, Cache, Queue, Rate Limiter)
│   │   ├── rabbitmq_broker.py # RabbitMQ 구현 (Direct, Fanout, Topic, DLQ, Priority, TTL)
│   │   └── kafka_broker.py    # Kafka 구현 (Basic, Keyed, Batch, Topic Mgmt)
│   ├── api/                   # FastAPI 라우트 (브로커별 분리)
│   └── monitoring/
│       ├── timer.py           # @measure_time 데코레이터 + Stopwatch (AOP)
│       └── metrics.py         # Prometheus 메트릭 (Histogram, Counter, Gauge)
└── notebooks/                 # 학습 노트북 (지금 여기!)
```

**핵심 설계**: 모든 브로커가 `AbstractBroker`를 상속하여 `connect()`, `disconnect()`, `publish()`, `subscribe()`, `benchmark()` 인터페이스를 통일합니다.

---
## 2. 환경 설정 (sys.path)

In [24]:
# 프로젝트 루트를 import 경로에 추가
# 왜? notebooks/ 폴더에서 실행하면 app 모듈을 찾을 수 없기 때문
import sys
from pathlib import Path

PROJECT_ROOT = str(Path.cwd().parent)
if PROJECT_ROOT not in sys.path:
    sys.path.insert(0, PROJECT_ROOT)

print(f"프로젝트 루트: {PROJECT_ROOT}")

프로젝트 루트: /Users/leebeanbin/Downloads/kafkaRabbitTest


In [25]:
# HTTP 클라이언트 준비 (FastAPI 서버 호출용)
import httpx

BASE_URL = "http://localhost:8000"
client = httpx.AsyncClient(base_url=BASE_URL, timeout=10.0)
print(f"API 클라이언트 준비 완료: {BASE_URL}")

API 클라이언트 준비 완료: http://localhost:8000


---
## 3. AbstractBroker - 공통 인터페이스 분석

`AbstractBroker`는 Python의 ABC(Abstract Base Class)를 사용하여 **모든 브로커가 반드시 구현해야 할 계약**을 정의합니다.

### 왜 ABC인가?
- SQLAlchemy는 DB ORM이지만, 여기서는 "서로 다른 브로커의 공통 행위"를 강제해야 합니다
- 이것은 **Strategy Pattern** 영역 → Python에서는 ABC로 해결
- Java의 `interface MessageBroker`와 동일한 역할

### 필수 구현 메서드
| 메서드 | 역할 | 반환 |
|--------|------|------|
| `name` | 브로커 고유 이름 (property) | `str` |
| `is_connected` | 연결 상태 (property) | `bool` |
| `connect()` | 브로커 연결 | `None` |
| `disconnect()` | 연결 종료 | `None` |
| `publish(destination, message)` | 메시지 발행 | `dict` |
| `subscribe(destination, callback)` | 메시지 구독 | `None` |
| `benchmark(destination, count)` | 벤치마크 실행 | `dict` |

In [26]:
# AbstractBroker 소스코드 직접 확인
import inspect
from app.brokers.base import AbstractBroker

# 추상 메서드 목록 추출
abstract_methods = [
    name for name, method in inspect.getmembers(AbstractBroker)
    if getattr(method, '__isabstractmethod__', False)
]
print("필수 구현 메서드:", abstract_methods)

필수 구현 메서드: ['benchmark', 'connect', 'disconnect', 'is_connected', 'name', 'publish', 'subscribe']


In [27]:
# ABC 강제력 테스트: 추상 메서드를 구현하지 않으면 TypeError 발생
try:
    class IncompleteBroker(AbstractBroker):
        pass  # 아무것도 구현하지 않음
    broker = IncompleteBroker()  # 여기서 에러!
except TypeError as e:
    print(f"예상대로 에러 발생: {e}")
    print("→ ABC가 누락된 구현을 컴파일 타임에 잡아줍니다")

예상대로 에러 발생: Can't instantiate abstract class IncompleteBroker without an implementation for abstract methods 'benchmark', 'connect', 'disconnect', 'is_connected', 'name', 'publish', 'subscribe'
→ ABC가 누락된 구현을 컴파일 타임에 잡아줍니다


### `destination` 통일 개념

`publish(destination, message)`에서 `destination`은 브로커마다 의미가 다릅니다:

| 브로커 | destination의 실체 | 예시 |
|--------|-------------------|------|
| Redis | channel 이름 | `"notifications"` |
| RabbitMQ | queue 이름 | `"order-queue"` |
| Kafka | topic 이름 | `"user-events"` |

이렇게 통일함으로써 `for broker in brokers: await broker.publish(dest, msg)` 루프가 가능합니다.

---
## 4. pydantic-settings 설정 분석

`pydantic-settings`는 환경변수를 **타입 안전한 Python 객체**로 변환합니다.
`.env` 파일이나 시스템 환경변수에서 자동으로 값을 읽어옵니다.

In [28]:
from app.config import Settings, settings

# 현재 설정 값 확인
print("=== 현재 설정 ===")
print(f"Redis URL:    {settings.redis_url}")
print(f"RabbitMQ URL: {settings.rabbitmq_url}")
print(f"Kafka:        {settings.kafka_bootstrap_servers}")
print(f"App:          {settings.app_host}:{settings.app_port}")

=== 현재 설정 ===
Redis URL:    redis://localhost:6379
RabbitMQ URL: amqp://guest:guest@localhost:5672/
Kafka:        localhost:9092
App:          0.0.0.0:8000


In [29]:
# pydantic-settings의 핵심: 환경변수 → 자동 매핑
# 예: REDIS_HOST=my-redis → settings.redis_host == "my-redis"
# docker-compose.yml에서 이 원리를 활용:
#   environment:
#     REDIS_HOST: redis        ← 컨테이너 이름
#     KAFKA_BOOTSTRAP_SERVERS: kafka:9094

print("\nmodel_config:", Settings.model_config)
print("→ env_file='.env'로 설정, 로컬에서는 .env 파일 없으면 기본값 사용")


model_config: {'extra': 'forbid', 'arbitrary_types_allowed': True, 'validate_default': True, 'case_sensitive': False, 'env_prefix': '', 'nested_model_default_partial_update': False, 'env_file': '.env', 'env_file_encoding': 'utf-8', 'env_ignore_empty': False, 'env_nested_delimiter': None, 'env_nested_max_split': None, 'env_parse_none_str': None, 'env_parse_enums': None, 'cli_prog_name': None, 'cli_parse_args': None, 'cli_parse_none_str': None, 'cli_hide_none_type': False, 'cli_avoid_json': False, 'cli_enforce_required': False, 'cli_use_class_docs_for_groups': False, 'cli_exit_on_error': True, 'cli_prefix': '', 'cli_flag_prefix_char': '-', 'cli_implicit_flags': False, 'cli_ignore_unknown_args': False, 'cli_kebab_case': False, 'cli_shortcuts': None, 'json_file': None, 'json_file_encoding': None, 'yaml_file': None, 'yaml_file_encoding': None, 'yaml_config_section': None, 'toml_file': None, 'secrets_dir': None, 'protected_namespaces': ('model_validate', 'model_dump', 'settings_customise_so

---
## 5. Pydantic 스키마 분석


`app/schemas.py`는 API 요청/응답의 데이터 검증을 담당합니다.
라우트와 분리하여 **단일 책임 원칙(SRP)**을 준수합니다.

In [30]:
from app.schemas import (
    MessageRequest, BenchmarkRequest, CacheRequest,
    TopicExchangeRequest, PriorityMessageRequest, KeyedMessageRequest
)

# 각 스키마의 필드와 기본값 확인
for schema in [MessageRequest, BenchmarkRequest, CacheRequest, KeyedMessageRequest]:
    fields = schema.model_fields
    print(f"\n{schema.__name__}:")
    for name, info in fields.items():
        default = info.default if info.default is not None else "(필수)"
        print(f"  {name}: {info.annotation.__name__ if hasattr(info.annotation, '__name__') else info.annotation} = {default}")


MessageRequest:
  content: str = PydanticUndefined
  metadata: dict = {}

BenchmarkRequest:
  message_count: int = 1000

CacheRequest:
  key: str = PydanticUndefined
  value: dict = {}
  ttl: int = 60

KeyedMessageRequest:
  key: str = PydanticUndefined
  content: str = PydanticUndefined
  metadata: dict = {}


In [31]:
# Pydantic 검증 테스트: 잘못된 값을 넣으면 에러
from pydantic import ValidationError

try:
    BenchmarkRequest(message_count=100000)  # 최대 50000
except ValidationError as e:
    print("검증 실패 (예상대로):")
    print(e.errors()[0]['msg'])

검증 실패 (예상대로):
Input should be less than or equal to 50000


---
## 6. 싱글턴 패턴과 Lifespan 관리

각 브로커 모듈 하단에 `redis_broker = RedisBroker()` 같은 **모듈 레벨 싱글턴**이 있습니다.
Python에서 모듈은 한 번만 로드되므로, 이것만으로 싱글턴이 보장됩니다.

### Lifespan 동작 흐름
```text
FastAPI 시작
  → lifespan() 진입
  → for broker in ALL_BROKERS:
      await broker.connect()    # Redis → RabbitMQ → Kafka 순서대로 연결
  → yield (서버 실행 중...)
  → for broker in reversed(ALL_BROKERS):
      await broker.disconnect() # Kafka → RabbitMQ → Redis 역순으로 종료
```

연결 실패해도 앱은 계속 동작합니다 (**graceful degradation**).

In [32]:
# 싱글턴 확인: 같은 객체를 참조하는지 검증
from app.brokers import redis_broker, rabbitmq_broker, kafka_broker
from app.lifespan import ALL_BROKERS

print("ALL_BROKERS에 등록된 브로커:")
for b in ALL_BROKERS:
    print(f"  {b!r}")

# 동일 객체 확인
assert redis_broker is ALL_BROKERS[0], "싱글턴이 아닙니다!"
print("\n✅ 싱글턴 확인 완료 - 모듈 import와 lifespan이 같은 객체를 공유")

ALL_BROKERS에 등록된 브로커:
  <RedisBroker name='redis' status=disconnected>
  <RabbitMQBroker name='rabbitmq' status=disconnected>
  <KafkaBroker name='kafka' status=disconnected>

✅ 싱글턴 확인 완료 - 모듈 import와 lifespan이 같은 객체를 공유


---
## 7. Docker 인프라 확인

`docker-compose.yml`에 정의된 서비스들:

| 서비스 | 이미지 | 포트 | 용도 |
|--------|--------|------|------|
| redis | redis:**8.0**-alpine | 6379 | 메시지 브로커 + 캐시 (JSON/Search/TimeSeries 내장) |
| rabbitmq | rabbitmq:**4**-management-alpine | 5672, **15672** | AMQP + Management UI (Quorum Queue 기본) |
| kafka | apache/kafka:**4.0** | 9092 | **KRaft 전용** (ZooKeeper 코드 완전 제거) |
| kafka-ui | ghcr.io/kafbat/kafka-ui | **8080** | Kafka 웹 관리 도구 |
| prometheus | prom/prometheus:v3.2.1 | **9090** | 메트릭 수집 (OpenTelemetry 네이티브) |
| grafana | grafana/grafana:11.5.2 | **3000** | 시각화 대시보드 |
| image-processor | Go 1.26 (빌드) | **8081** | 이미지 처리 마이크로서비스 |
| app | Python 3.13 (빌드) | **8000** | FastAPI 앱 (Swagger: /docs) |

**왜 이 버전들인가?**
- **Redis 8**: Redis Stack 모듈(JSON, Search, TimeSeries, Bloom)이 코어에 통합. 별도 설치 불필요
- **RabbitMQ 4**: Quorum Queue가 기본 복제 큐 타입. Classic Mirrored Queue 제거로 단순화
- **Kafka 4.0**: ZooKeeper 코드가 완전히 삭제됨. KRaft가 유일한 메타데이터 관리 방식

In [33]:
# 헬스체크: 각 서비스가 정상인지 확인
import httpx

health_checks = {
    "FastAPI": f"{BASE_URL}/health",
    "RabbitMQ Management": "http://localhost:15672",
    "Kafka UI": "http://localhost:8080",
    "Prometheus": "http://localhost:9090/-/healthy",
}

for name, url in health_checks.items():
    try:
        r = httpx.get(url, timeout=3.0)
        status = "✅" if r.status_code < 400 else f"⚠️ {r.status_code}"
    except Exception as e:
        status = f"❌ {type(e).__name__}"
    print(f"{status} {name}: {url}")

✅ FastAPI: http://localhost:8000/health
✅ RabbitMQ Management: http://localhost:15672
✅ Kafka UI: http://localhost:8080
✅ Prometheus: http://localhost:9090/-/healthy


In [34]:
# FastAPI /health 엔드포인트 상세 결과
r = await client.get("/health")
health = r.json()

print("=== 브로커 헬스체크 ===")
for broker_info in health.get("brokers", []):
    name = broker_info.get("broker", "unknown")
    connected = broker_info.get("connected", False)
    icon = "✅" if connected else "❌"
    print(f"  {icon} {name}: connected={connected}")

=== 브로커 헬스체크 ===


---
## 8. 첫 메시지 발행 - 각 브로커 테스트

FastAPI API를 통해 각 브로커에 첫 번째 메시지를 보내봅니다.
모든 브로커가 동일한 요청 형태(`MessageRequest`)를 받습니다.

In [35]:
# Redis Pub/Sub - 첫 메시지
r = await client.post("/redis/pubsub/publish", json={
    "content": "Hello from Jupyter!",
    "metadata": {"source": "notebook", "notebook": "01"}
})
print("Redis Pub/Sub 결과:")
print(r.json())

Redis Pub/Sub 결과:
{'broker': 'redis', 'pattern': 'pubsub', 'destination': 'test-channel', 'subscribers_received': 0, 'elapsed_ms': 2.139}


In [36]:
# RabbitMQ Direct - 첫 메시지
r = await client.post("/rabbitmq/direct/publish",
    params={"queue_name": "order-queue"},
    json={
        "content": "Hello from Jupyter!",
        "metadata": {"source": "notebook", "notebook": "01"}
    }
)
print("RabbitMQ Direct 결과:")
print(r.json())

RabbitMQ Direct 결과:
{'broker': 'rabbitmq', 'pattern': 'direct', 'destination': 'test-queue', 'queue_message_count': 2, 'elapsed_ms': 12.821}


In [37]:
# Kafka Basic - 첫 메시지
r = await client.post("/kafka/basic/publish", json={
    "content": "Hello from Jupyter!",
    "metadata": {"source": "notebook", "notebook": "01"}
})
print("Kafka Basic 결과:")
result = r.json()
print(result)
print(f"\n→ partition={result.get('partition')}, offset={result.get('offset')}")

Kafka Basic 결과:
{'broker': 'kafka', 'pattern': 'basic', 'destination': 'test-topic', 'partition': 0, 'offset': 3, 'elapsed_ms': 50.317}

→ partition=0, offset=3


### 응답 비교

세 브로커의 응답에서 공통점과 차이점을 확인합니다:

| 필드 | Redis | RabbitMQ | Kafka |
|------|-------|----------|-------|
| `broker` | ✅ | ✅ | ✅ |
| `pattern` | pubsub | direct | basic |
| `destination` | ✅ | ✅ | ✅ |
| `elapsed_ms` | ✅ | ✅ | ✅ |
| 고유 필드 | subscribers_received | queue_message_count | partition, offset |

→ `@measure_time` 데코레이터가 `elapsed_ms`를 자동 주입합니다 (AOP 패턴)

---
## 9. Swagger UI 활용법

브라우저에서 `http://localhost:8000/docs`를 열면 Swagger UI를 통해
모든 API를 인터랙티브하게 테스트할 수 있습니다.

### 주요 엔드포인트 그룹
- `/redis/*` - Redis 5가지 패턴 (Pub/Sub, Stream, Queue, Cache, Rate Limiter)
- `/rabbitmq/*` - RabbitMQ 6가지 패턴 (Direct, Fanout, Topic, DLQ, Priority, TTL)
- `/kafka/*` - Kafka 3가지 패턴 (Basic, Keyed, Batch) + Topic 관리
- `/benchmark/*` - 성능 벤치마크 (개별 + 비교)
- `/monitoring/*` - Prometheus 메트릭, 발행 히스토리
- `/health` - 브로커 헬스체크

In [38]:
# API 라우트 목록 확인 (OpenAPI spec에서 추출)
r = await client.get("/openapi.json")
paths = r.json()["paths"]

print(f"총 {len(paths)}개 엔드포인트:\n")
for path, methods in sorted(paths.items()):
    for method in methods:
        if method in ("get", "post", "put", "delete"):
            summary = methods[method].get("summary", "")
            print(f"  {method.upper():6s} {path:45s} {summary}")

총 40개 엔드포인트:

  POST   /api/direct                                   Direct Api
  POST   /benchmark/all                                Benchmark All
  POST   /benchmark/kafka                              Benchmark Kafka
  POST   /benchmark/kafka-batch                        Benchmark Kafka Batch
  POST   /benchmark/rabbitmq                           Benchmark Rabbitmq
  POST   /benchmark/redis                              Benchmark Redis
  POST   /benchmark/redis-stream                       Benchmark Redis Stream
  GET    /health                                       Health Check
  POST   /kafka/basic/publish                          Basic Publish
  POST   /kafka/batch/publish                          Batch Publish
  POST   /kafka/keyed/publish                          Keyed Publish
  POST   /kafka/topic/create                           Topic Create
  GET    /kafka/topic/info/{topic}                     Topic Info
  GET    /kafka/topics                                 List Topics
  GE

---
## 🏁 10. 정리 - 핵심 포인트

### 📌 핵심 개념 요약

| 개념 | 설명 |
|------|------|
| **ABC 패턴** | 모든 브로커가 동일한 인터페이스를 구현하도록 강제 |
| **destination 통일** | Redis channel, RabbitMQ queue, Kafka topic → 모두 `destination` |
| **pydantic-settings** | 환경변수를 타입 안전한 객체로 자동 변환 |
| **모듈 싱글턴** | `redis_broker = RedisBroker()` → import 시 동일 인스턴스 공유 |
| **Lifespan** | 앱 시작/종료 시 브로커 연결을 자동 관리 |
| **@measure_time** | AOP 데코레이터로 elapsed_ms 자동 주입 + Prometheus 기록 |

> **📝 Note**: 이 노트북에서 다룬 개념들은 이후 모든 노트북의 기반이 됩니다.
> ABC 패턴과 destination 통일을 이해해야 각 브로커 심화 노트북의 코드가 자연스럽게 읽힙니다.

---

### 🗺️ 다음 단계 바로가기

| 노트북 | 내용 | 난이도 |
|--------|------|--------|
| [02_redis_deep_dive.ipynb](./02_redis_deep_dive.ipynb) | Redis 5가지 패턴 (Pub/Sub, Stream, Queue, Cache, Rate Limiter) | ⭐⭐ |
| [03_rabbitmq_deep_dive.ipynb](./03_rabbitmq_deep_dive.ipynb) | RabbitMQ AMQP 프로토콜 심화 (Exchange, DLQ, Priority) | ⭐⭐ |
| [04_kafka_deep_dive.ipynb](./04_kafka_deep_dive.ipynb) | Kafka 분산 아키텍처 심화 (Partition, Consumer Group, Batch) | ⭐⭐ |

> **💡 Tip**: 02~04번은 서로 독립적이므로 **관심 있는 브로커부터** 학습해도 됩니다.

### 🌐 유용한 웹 UI 링크
- 📊 **Swagger UI**: http://localhost:8000/docs
- 🐰 **RabbitMQ Management**: http://localhost:15672 (guest/guest)
- 📈 **Kafka UI**: http://localhost:8080
- 📉 **Prometheus**: http://localhost:9090
- 📊 **Grafana**: http://localhost:3000 (admin/admin)

In [39]:
# 정리
await client.aclose()
print("✅ 노트북 01 완료!")

✅ 노트북 01 완료!
